### ISSUE TITLE
Stream Docker image tarball to disk instead of memory dirst

### ISSUE THREAD

**Issue by cstrzadala** [docker/docker-py#1352]
I'm attempting to create a tarball of an existing Docker image via the https://docker-py.readthedocs.io/en/stable/images.html save example. When attempting to write the data to disk, I get `MemoryError` during the writing process.

The image is 1.516GB so I'm assuming we are running out of memory (this process is within a container).

Can the save be done directly to disk and avoid being pushed into memory?

**Comment by shin-** [comment]
We should provide a streaming option. You can try the following code excerpt in the meantime:

```python
import docker

client = docker.from_env()
img = client.images.get('busybox:latest')
url = client.api._url('/images/{0}/get', img.id)
res = client.api._get(url, stream=True)
with open('/tmp/img.tar', 'w') as f:
  for chunk in res.iter_content(chunk_size=None):
    f.write(chunk)
```

**Comment by cstrzadala** [comment]
Thanks @shin-, that worked a treat. When I got sometime, I'll do a PR to add this into the project.

**Comment by realityone** [comment]
Actually docker-py already provide the streaming way to save an image.

```python
image = cli.images.get("fedora:latest")
resp = image.save()
f = open('/tmp/fedora-latest.tar', 'w')
for chunk in resp.stream():
    f.write(chunk)
f.close()
```

I updated the docstring for this https://github.com/docker/docker-py/pull/1377.

**Comment by inge5t0r** [comment]
@realityone 

This doesn't have the desired effect. The entire image is still read into memory with `resp = image.save()`.

**Comment by realityone** [comment]
@inge5t0r 
I remember the image.save() just send the request and will return a raw response which has not read yet.

If you mean this request will block sometime, it may be caused by docker daemon, docker daemon is packing the image.

**Comment by david-drinn** [comment]
In docker-py 3.0.0, `image.save()` returns a generator object:

    In [73]: image.save()
    Out[73]: <generator object APIClient._stream_raw_result at 0x7fc8292d0780>

If I use this as the [3.0.0 docs implies](http://docker-py.readthedocs.io/en/3.0.0/images.html#image-objects) (I say imply, since it actually uses `cli.get_image()` and not `image.save()`), it takes several minutes to write a 30 MB docker image to tarball file.

The lower level workaround is to use the same @shin- code snippet from above, which takes less than a second to write the same image to tarball file:

    import docker
    
    client = docker.from_env()
    img = client.images.get('busybox:latest')
    url = client.api._url('/images/{0}/get', img.id)
    res = client.api._get(url, stream=True)
    with open('/tmp/img.tar', 'w') as f:
      for chunk in res.iter_content(chunk_size=None):
        f.write(chunk)

I could use some clarity here on whether I'm using `image.save()` incorrectly.

**Comment by realityone** [comment]
The image save example in docker-py is iterating the `urllib3.response.HTTPResponse` directly, which is inherited from `io.IOBase`. In this way, every chunk is separated by `\x0a`, so the chunk size is not guaranteed. This will take several time to read the whole response.

I prefer using `resp.stream(amt=***)` to iterate the response stream, specify the chunk size manually, this will save a lot of time to read the response.

**Comment by david-drinn** [comment]
3.0.0 I think must give a different object type from `image.save()` than previous versions, since it doesn't have the `stream()` attribute.

```python
In [96]: resp = image.save()

In [97]: resp.stream
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-97-0cd3e4d9f8f0> in <module>()
----> 1 resp.stream

AttributeError: 'generator' object has no attribute 'stream'
```

**Comment by realityone** [comment]
Hi, @david-drinn 
After digging further, I found that in docker-py 3.0.*, the `image.save()` is using `_stream_raw_result` to iterate the response. The chunk size in `_stream_raw_result` method is 1. This will take a lot of time to read the whole response.

Hi @shin- ,
Maybe we should not using `_stream_raw_result` in `image.save()`, or the `_stream_raw_result` method should support customizing the chunk size.

Which one do you prefer, I will create an PR to fix this issue.

**Comment by goldmann** [comment]
I agree with @realityone, setting chunk size to 1 is a bad idea IMHO. We deal with images that weight few GB. This is a regression for me and I've opened: #1905 to track this specific issue.

**Comment by shin-** [comment]
Just wanted to say - I read the messages, and it's on my radar - should be a simple fix. Thanks everyone for the reports.
